crates/sdk/src/masp.rs

```diff
@@ -609,10 +657,81 @@ pub struct ShieldedContext<U: ShieldedUtils> {
     pub tx_note_map: BTreeMap<IndexedTx, usize>,
     /// A cache of fetched indexed txs.
     pub unscanned: Unscanned,
+    /// We cannot update spent notes until all fetched notes have been
+    /// decrypted. This temporarily stores the relevant encrypted data in
+    /// case syncing is interrupted.
+    pub decrypted_note_cache: HashMap<(IndexedTx, ViewingKey), DecryptedData>,
     /// The sync state of the context
     pub sync_status: ContextSyncStatus,
 }
 
+/// A thread safe handle back to a shielded context
+#[derive(Clone)]
+pub struct ThreadHandle<'shielded, U: ShieldedUtils + MaybeSend + MaybeSync> {
+    // for safety, this field should not be accessible from the outside
+    inner: Arc<Mutex<&'shielded mut ShieldedContext<U>>>,
+}
+
+unsafe impl<'shielded, U: ShieldedUtils + MaybeSend + MaybeSync> Send
+    for ThreadHandle<'shielded, U>
+{
+}
+unsafe impl<'shielded, U: ShieldedUtils + MaybeSend + MaybeSync> Sync
+    for ThreadHandle<'shielded, U>
+{
+}
```

Comment:
this might be the source of a nasty bug if `ShieldedUtils` is `!Send` or `!Sync` I think, no?

Reply:
The main reason, this kind of thing isn't send or sync is because you could copy the &mut, or take the inner value. As long we never do that, this really shouldn't be an issue I think. The inner pointer is private and this thing should only be interacted with via the type api.

Reply:
see https://github.com/anoma/namada/pull/2777#discussion_r1507568036